tools/ac_harness/racing_telemetry.py

```
@@ -0,0 +1,171 @@
+
+def record(out_path: str, *, max_laps: int = 10, max_seconds: float = 1800.0) -> int:
+    """Record physics+graphics frames to ``out_path`` (CSV) until ``max_laps`` or ``max_seconds``.
+
+    Windows/rig-only. Dedupes on the physics packetId, segments laps on ``completedLaps``, and
+    prints a per-lap summary (time, min/max speed). Returns the number of completed laps.
+    """
+    if sys.platform != "win32":
+        raise RuntimeError("racing_telemetry recording is Windows-only (AC shared memory)")
+    phys = open_shared_memory(SHM_PHYSICS, _PHYS_BYTES)
+    gfx = open_shared_memory(SHM_GRAPHICS, _GFX_BYTES)
+    fh = open(out_path, "w", encoding="utf-8", newline="\n")
```

Ruff (0.15.17)

[warning] Use a context manager for opening files

(SIM115)